zerver/lib/request.py

```diff
@@ -350,7 +384,18 @@ def _wrapped_view_func(request: HttpRequest, *args: Any, **kwargs: Any) -> HttpR
                 except Exception:
                     raise JsonableError(_('Argument "{}" is not valid JSON.').format(post_var_name))
 
-                error = param.validator(post_var_name, val)
+                # TODO: set these up when has_request_variables
+                #       is called, rather than every time
+                if param.validator.__name__ == 'auto':
+                    # TODO: break circular deps so we can move this to
+                    #       the top with other imports
+                    from zerver.lib.validator import build_auto_validator
+                    var_type = type_hints[func_var_name]
+                    validator = build_auto_validator(var_type)
+                else:
+                    validator = param.validator
```

Comment:
We likely don't want to be doing this logic every time we call the function.

Reply:
hence this comment:

~~~
397                 # TODO: set these up when has_request_variables
398                 #       is called, rather than every time
~~~

That last commit is still rough, but it's easy to move this logic up.